packages-exp/functions-exp/src/service.ts

```diff
@@ -0,0 +1,266 @@
+        return resolve();
+      };
+    });
+  }
+
+  get app(): FirebaseApp {
+    return this.app_;
+  }
+
+  setRegion(region: string): void {
+    this.region_ = region;
+  }
+
+  INTERNAL = {
+    delete: (): Promise<void> => {
+      return this.deleteService();
+    }
+  };
+
+  /**
+   * Returns the URL for a callable with the given name.
+   * @param name The name of the callable.
+   */
+  _url(name: string): string {
+    const projectId = this.app_.options.projectId;
+    const region = this.region_;
+    if (this.emulatorOrigin !== null) {
+      const origin = this.emulatorOrigin;
+      return `${origin}/${projectId}/${region}/${name}`;
+    }
+    return `https://${region}-${projectId}.cloudfunctions.net/${name}`;
+  }
+
+  /**
+   * Changes this instance to point to a Cloud Functions emulator running
+   * locally. See https://firebase.google.com/docs/functions/local-emulator
+   *
+   * @param origin The origin of the local emulator, such as
+   * "http://localhost:5005".
+   */
+  useFunctionsEmulator(origin: string): void {
+    this.emulatorOrigin = origin;
+  }
+
+  /**
+   * Returns a reference to the callable https trigger with the given name.
+   * @param name The name of the trigger.
+   */
+  httpsCallable(name: string, options?: HttpsCallableOptions): HttpsCallable {
+    return data => {
+      return this.call(name, data, options || {});
+    };
+  }
+
+  /**
+   * Does an HTTP POST and returns the completed response.
+   * @param url The url to post to.
+   * @param body The JSON body of the post.
+   * @param headers The HTTP headers to include in the request.
+   * @return A Promise that will succeed when the request finishes.
+   */
+  private async postJSON(
+    url: string,
+    body: {},
+    headers: Headers
+  ): Promise<HttpResponse> {
+    headers.append('Content-Type', 'application/json');
+
+    let response: Response;
+    try {
+      response = await fetch(url, {
+        method: 'POST',
+        body: JSON.stringify(body),
+        headers
+      });
+    } catch (e) {
+      // This could be an unhandled error on the backend, or it could be a
+      // network error. There's no way to know, since an unhandled error on the
+      // backend will fail to set the proper CORS header, and thus will be
+      // treated as a network error by fetch.
+      return {
+        status: 0,
+        json: null
+      };
+    }
+    let json: {} | null = null;
+    try {
+      json = await response.json();
+    } catch (e) {
+      // If we fail to parse JSON, it will fail the same as an empty body.
+    }
+    return {
+      status: response.status,
+      json
+    };
+  }
+
+  /**
+   * Calls a callable function asynchronously and returns the result.
+   * @param name The name of the callable trigger.
+   * @param data The data to pass as params to the function.s
+   */
+  private async call(
+    name: string,
+    data: unknown,
+    options: HttpsCallableOptions
+  ): Promise<HttpsCallableResult> {
```

Comment:
These functions should be moved outside of the Class to be top level functions and take an instance of the Class as a parameter to access context.

Currently it will not make much difference in terms of import size because `httpsCallable` is the entire API surface, but it will set up the structure that allows us to add functionalities in a modular way in the future.

Reply:
I'm not totally sure I did it correctly, take a look.